src/borg/_hashindex.c

```diff
@@ -248,99 +262,130 @@ count_empty(HashIndex *index)
 }
 
 /* Public API */
+
 static HashIndex *
-hashindex_read(const char *path)
+hashindex_read(PyObject *file)
 {
-    FILE *fd;
-    off_t length, buckets_length, bytes_read;
-    HashHeader header;
+    Py_ssize_t length, buckets_length, bytes_read;
+    HashHeader *header;
     HashIndex *index = NULL;
 
-    if((fd = fopen(path, "rb")) == NULL) {
-        EPRINTF_PATH(path, "fopen for reading failed");
-        return NULL;
-    }
-    bytes_read = fread(&header, 1, sizeof(HashHeader), fd);
-    if(bytes_read != sizeof(HashHeader)) {
-        if(ferror(fd)) {
-            EPRINTF_PATH(path, "fread header failed (expected %ju, got %ju)",
-                         (uintmax_t) sizeof(HashHeader), (uintmax_t) bytes_read);
-        }
-        else {
-            EPRINTF_MSG_PATH(path, "fread header failed (expected %ju, got %ju)",
-                             (uintmax_t) sizeof(HashHeader), (uintmax_t) bytes_read);
-        }
-        goto fail;
+    PyObject *header_bytes = PyObject_CallMethod(file, "read", "n", sizeof(HashHeader));
+    if(!header_bytes) {
+        /* Python exception occurred; just return */
+        return 0;
     }
-    if(fseek(fd, 0, SEEK_END) < 0) {
-        EPRINTF_PATH(path, "fseek failed");
-        goto fail;
+    bytes_read = PyBytes_Size(header_bytes);
+    if(PyErr_Occurred()) {
+        /* TypeError, not a bytes() object */
+        goto fail_decref_header;
     }
-    if((length = ftell(fd)) < 0) {
-        EPRINTF_PATH(path, "ftell failed");
-        goto fail;
+    if(bytes_read != sizeof(HashHeader)) {
+        /* Truncated file */
+        /* Note: %zd is the format for Py_ssize_t */
+        PyErr_Format(PyExc_ValueError, "Could not read header (expected %zd, but read %zd bytes)", sizeof(HashHeader), bytes_read);
+        goto fail_decref_header;
     }
-    if(fseek(fd, sizeof(HashHeader), SEEK_SET) < 0) {
-        EPRINTF_PATH(path, "fseek failed");
-        goto fail;
+
+    /* Find length of file */
+    PyObject *length_object = PyObject_CallMethod(file, "seek", "ni", 0, SEEK_END);
```

Comment:
maybe it would be a nice convention for *.c to always have _obj suffix if it is a python object.